File: 113.py
# Definition for a binary tree node.
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def visiteNode(self, targetSum, subtotal, node, foundsPaths, currentPath):
        newCurrentPath = currentPath.copy()
        newCurrentPath.append(node.val)
        if node.left != None:
            self.visiteNode(targetSum, subtotal + node.val, node.left, foundsPaths, newCurrentPath)

        if node.right != None:
            self.visiteNode(targetSum, subtotal + node.val, node.right, foundsPaths, newCurrentPath)

        if node.left == None and node.right == None:
            if node.val + subtotal == targetSum:
                foundsPaths.append(newCurrentPath)
                


    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        if root == None:
            return []

        found_paths = []

        self.visiteNode(targetSum, 0, root, found_paths, [])
        return found_paths

solution = Solution()
print(solution.pathSum(TreeNode(1, TreeNode(3), TreeNode(2)), 4))

File: 113.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function pathSum(root: TreeNode | null, targetSum: number): number[][] {
  const res: number[][] = [];

  const dfs = (
    node: TreeNode | null,
    ancestors: number[],
    targetSum: number,
  ) => {
    if (!node) {
      return;
    }

    const currentPath = [...ancestors, node.val];
    const currentTargetSum = targetSum - node.val;

    if (!node.left && !node.right && currentTargetSum == 0) {
      res.push(currentPath);
      return;
    }

    dfs(node.left, currentPath, currentTargetSum);
    dfs(node.right, currentPath, currentTargetSum);
  };

  dfs(root, [], targetSum);

  return res;
}
